README.md
[![codecov.io](https://codecov.io/github/mozilla-services/push-messages/coverage.svg?branch=master)](https://codecov.io/github/mozilla-services/push-messages?branch=master) [![Build Status](https://travis-ci.org/mozilla-services/push-messages.svg?branch=master)](https://travis-ci.org/mozilla-services/push-messages)

# Push Messages API

A [swagger](http://swagger.io/) driven Pyramid app that utilizes Redis for
recent messages sent into the Mozilla Push System and manages crypto public keys
that should have their recent message history recorded.

## Architecture

This project provides an internal API for the Mozilla Push Developer Dashboard.
Developers are identified by their registered VAPID public keys. A separate
process populates those keys to a Redis database.

### DynamoDB Table

Stores a permanent list of VAPID public keys that should have their message
metadata extracted from the Mozilla Push Service logstream.

This is a very lightly used table managed by `push_messages`.

### Redis

Stores a circular list of recent message metadata, indexed by VAPID public key.
`push_messages` can retrieve this data via API calls. Redis supports efficient
access/retrieval and has trim operations to emulate a circular list structure
efficiently.

## Developing

Checkout this repo, you will need Redis installed and a local DynamoDB to test
against.

Then:

    $ virtualenv pmenv
    $ source pmenv/bin/activate
    $ pip install -r requirements.txt
    $ python setup.py develop

Edit the `development.ini` to point to your Redis database and set the DynamoDB
table name to use.

Run the app locally:

    $ pserve development.ini

It will then let you know where you can reach it.


CHANGELOG.md
<a name="0.7"></a>
## 0.7 (2016-04-14)


#### Bug Fixes

*   heartbeat shouldn't check for keytable ([45080730](https://github.com/mozilla-services/push-messages/commit/45080730f54cb06e4ea68e7dce2f7e53b8dee823))



<a name="0.6"></a>
## 0.6 (2016-04-12)


#### Features

*   add circle.yml file for Dockerflow ([f125d53b](https://github.com/mozilla-services/push-messages/commit/f125d53beb3fb77703c481ec0fa2b5f004763928), closes [#14](https://github.com/mozilla-services/push-messages/issues/14))

#### Bug Fixes

*   properly return no content for 204 ([49d6ede1](https://github.com/mozilla-services/push-messages/commit/49d6ede19c0db1eb6cdbcaf6d4b1cb77860846da), closes [#16](https://github.com/mozilla-services/push-messages/issues/16))



<a name="0.5"></a>
## 0.5 (2016-04-11)


#### Features

*   return 404 for no registered key, and refactor for redis only ([959d20a6](https://github.com/mozilla-services/push-messages/commit/959d20a61b09a6e7272923feecd42b156fd987cd), closes [#6](https://github.com/mozilla-services/push-messages/issues/6))



<a name="0.4"></a>
## 0.4 (2016-04-08)


#### Bug Fixes

*   coerce message values to ensure validation passes ([01c5d40b](https://github.com/mozilla-services/push-messages/commit/01c5d40b4cf06e1ff97773ac26b6e3e86effe799))



<a name="0.3"></a>
## 0.3 (2016-04-07)


#### Bug Fixes

*   use floats for e-notation value ([47327494](https://github.com/mozilla-services/push-messages/commit/47327494d63505cc689174c2c9232a21828e4a2c))



<a name="0.2"></a>
## 0.2 (2016-04-07)


#### Bug Fixes

*   timestamp should be a integer ([b7ed0448](https://github.com/mozilla-services/push-messages/commit/b7ed04488ad0d0a799c3596d7081ee87443d15f3))



<a name="0.1"></a>
## 0.1 (2016-04-06)


#### Bug Fixes

*   restrict scan to avoid tests import ([03991fd7](https://github.com/mozilla-services/push-messages/commit/03991fd77ef92b9a6a22515c2afde0afe79ca8f9))

#### Test

*   add full test suite for unit tests and wsgi app ([8a6bad72](https://github.com/mozilla-services/push-messages/commit/8a6bad726a95bf5ee4647806f2566a4b1156bbec))

#### Chore

*   add clog toml for changelog gen ([d93b3db1](https://github.com/mozilla-services/push-messages/commit/d93b3db164c227f7fa1d51efdd0021424d10e458))
*   add initial project files and project base ([5c048d32](https://github.com/mozilla-services/push-messages/commit/5c048d32cc4fd2ca97d38a473700fc85f591513d))

#### Doc

*   add architecture background ([ce98cd98](https://github.com/mozilla-services/push-messages/commit/ce98cd98398c49077b522b6d81769fd601773cc9))

#### Features

*   add dockerized building per Dockerflow specs ([3728713f](https://github.com/mozilla-services/push-messages/commit/3728713fafa6e7d78963be658215f460d3ae0259), closes [#8](https://github.com/mozilla-services/push-messages/issues/8))
*   initial working push messages api ([691386b4](https://github.com/mozilla-services/push-messages/commit/691386b42d034ad6513dedb2f81d3a35548caaf7))


push_messages/tests.py
import json
import logging
import time
import uuid
import unittest

from mock import Mock, patch
from nose.tools import eq_, raises
from pyramid import testing
from pyramid.exceptions import NotFound


log = logging.getLogger(__name__)


class ViewTests(unittest.TestCase):
    def setUp(self):
        self.config = testing.setUp()

    def tearDown(self):
        testing.tearDown()

    def test_wsgi_app(self):
        from push_messages import main
        from webtest import TestApp
        app = main({}, **{
            "redis_host": "localhost",
            "redis_db": 0,
            "pyramid_swagger.schema_directory": ".",
            "pyramid_swagger.schema_file": "push_api.yaml",
            "debug": "true",
        })
        app.registry.redis_server.delete("registered_keys")
        testapp = TestApp(app)
        result = testapp.get("/keys", status=200)
        data = json.loads(result.body)
        eq_(data["keys"], [])

        result = testapp.get("/messages/fred", status=404)
        eq_(result.body, "")

        # Insert a key
        msg = dict(
            id="asdfasdf",
            ttl="60",
            timestamp=time.time(),
            size=20
        )
        dbkey = uuid.uuid4().hex
        app.registry.redis_server.hset("registered_keys", dbkey, "")
        app.registry.redis_server.lpush(dbkey, json.dumps(msg))

        result = testapp.get("/messages/%s" % dbkey)
        eq_(json.loads(result.body), dict(
            messages=[dict(
                id="asdfasdf",
                ttl=60,
                timestamp=msg["timestamp"],
                size=20
            )]
        ))

        # Verify 204
        app.registry.redis_server.delete(dbkey)
        result = testapp.get("/messages/%s" % dbkey, status=204)
        eq_(result.body, "")

    @patch("push_messages.resolve_elasticache_node")
    def test_wsgi_app_with_elasticache(self, mock_resolve):
        from push_messages import main
        from webtest import TestApp
        mock_resolve.return_value = "localhost"
        app = main({}, **{
            "redis_elasticache": "acachename",
            "redis_db": 0,
            "pyramid_swagger.schema_directory": ".",
            "pyramid_swagger.schema_file": "push_api.yaml",
            "debug": "true",
        })
        app.registry.redis_server.delete("registered_keys")
        testapp = TestApp(app)
        result = testapp.get("/keys", status=200)
        data = json.loads(result.body)
        eq_(data["keys"], [])
        eq_(len(mock_resolve.mock_calls), 1)

    def test_get_keys(self):
        from .views import get_keys
        request = testing.DummyRequest()
        request.redis = Mock()
        request.redis.hkeys.return_value = ["asdf"]
        request.key_table = Mock()
        request.key_table.all_keys.return_value = [
            dict(pubkey="asdf")
        ]
        info = get_keys(request)
        eq_(info["keys"][0]["public-key"], "asdf")

    def test_post_key(self):
        from .views import register_key
        request = testing.DummyRequest()
        request.redis = Mock()
        request.swagger_data = dict(key={"public-key": "asdf"})
        info = register_key(request)
        eq_(info.status_code, 201)
        request.redis.hset.assert_called_with("registered_keys", "asdf", "")

    def test_delete_key(self):
        from .views import delete_key
        request = testing.DummyRequest()
        request.redis = Mock()
        request.matchdict = dict(key="asdf")
        info = delete_key(request)
        eq_(info.status_code, 204)
        request.redis.hdel.assert_called_with("registered_keys", "asdf")

    def test_get_messages(self):
        from .views import get_messages
        request = testing.DummyRequest()
        request.matchdict["key"] = "something"
        request.redis = Mock()
        request.redis.hexists.return_value = True
        request.redis.lrange.return_value = [
            json.dumps(dict(id=2, timestamp=7894721893,
                            size=313, ttl=200))
        ]
        info = get_messages(request)
        eq_(len(info["messages"]), 1)
        request.redis.hexists.assert_called_with("registered_keys",
                                                 "something")

    @raises(NotFound)
    def test_get_messages_404_error(self):
        from .views import get_messages
        request = testing.DummyRequest()
        request.matchdict["key"] = "something"
        request.redis = Mock()
        request.redis.hexists.return_value = False
        get_messages(request)

    def test_get_messages_204(self):
        from .views import get_messages
        request = testing.DummyRequest()
        request.matchdict["key"] = "something"
        request.redis = Mock()
        request.redis.hexists.return_value = True
        request.redis.lrange.return_value = []
        response = get_messages(request)
        eq_(response.status_code, 204)
        request.redis.hexists.assert_called_with("registered_keys",
                                                 "something")

    def test_version(self):
        from .views import version
        request = testing.DummyRequest()
        info = version(request)
        eq_(len(info), 3)

    def test_heartbeat(self):
        from .views import heartbeat
        request = testing.DummyRequest()
        request.redis = Mock()
        info = heartbeat(request)
        eq_(len(request.redis.mock_calls), 1)
        eq_(info, {})

    def test_lbheartbeat(self):
        from .views import lbheartbeat
        request = testing.DummyRequest()
        info = lbheartbeat(request)
        eq_(info, {})


class DbTests(unittest.TestCase):
    @patch("push_messages.db.boto3")
    def test_resolve_elasticache(self, mock_boto):
        from push_messages.db import resolve_elasticache_node
        mock_boto.client.return_value = mock_client = Mock()
        mock_client.describe_cache_clusters.return_value = dict(
            CacheClusters=[
                dict(
                    CacheNodes=[
                        dict(
                            Endpoint=dict(Address="localhost")
                        )
                    ]
                )
            ]
        )
        resolve_elasticache_node("somename")
        eq_(len(mock_client.mock_calls), 1)
        eq_(len(mock_boto.mock_calls), 2)

        # No cluster
        mock_client.describe_cache_clusters.return_value = dict(
            CacheClusters=[]
        )

        @raises(Exception)
        def testit():
            resolve_elasticache_node("somename")
        testit()


app.py
from wsgiref.simple_server import make_server
from pyramid.config import Configurator
from pyramid.response import Response


APP_SETTINGS = {
    'pyramid_swagger.schema_directory': ".",
    'pyramid_swagger.schema_file': "push_api.yaml"
}


def get_keys(request):
    return dict(keys=[])


def get_messages(request):
    key = request.matchdict['key']
    return dict(messages=[])


def main(global_conf, **settings):
    config = Configurator(settings=settings)
    config.include('pyramid_swagger')
    config.add_route('get_keys', '/keys', request_method='GET')
    config.add_route('get_messages', '/messages/{key}', request_method='GET')

    config.add_view(get_keys, route_name='get_keys', renderer='json')
    config.add_view(get_messages, route_name='get_messages', renderer='json')
    return config.make_wsgi_app()


if __name__ == '__main__':
    app = main({}, **APP_SETTINGS)
    server = make_server('0.0.0.0', 8080, app)
    server.serve_forever()


push_messages/__init__.py
import os

import redis
from pyramid.config import Configurator

from push_messages.db import resolve_elasticache_node


__version__ = "0.7"


def main(global_config, **settings):
    """ This function returns a Pyramid WSGI application.
    """
    # Load a default aws default region
    if "AWS_DEFAULT_REGION" not in os.environ:
        os.environ["AWS_DEFAULT_REGION"] = "us-east-1"

    # Env overrides
    elasticache = os.environ.get("REDIS_ELASTICACHE",
                                 settings.get("redis_elasticache"))
    if elasticache:
        redis_host = resolve_elasticache_node(elasticache)
    else:
        redis_host = os.environ.get("REDIS_HOST", settings['redis_host'])
    redis_port = int(os.environ.get("REDIS_PORT", "6379"))
    redis_db = int(os.environ.get("REDIS_DB", settings['redis_db']))

    redis_server = redis.StrictRedis(
        host=redis_host,
        port=redis_port,
        db=redis_db,
    )

    settings["pyramid_swagger.exclude_paths"] = [
        r"^/__version__",
        r"^/__heartbeat__",
        r"^/__lbheartbeat__",
    ]

    # Provide a redis server instance off the request
    def rs(request):
        return redis_server

    if 'debug' in settings:
        settings['pyramid_swagger.exclude_routes'] = ['debugtoolbar']
    config = Configurator(settings=settings)
    config.registry.redis_server = redis_server
    config.add_request_method(rs, 'redis', reify=True)

    config.include('pyramid_swagger')

    config.add_route('get_keys', '/keys', request_method='GET')
    config.add_route('register_key', '/keys', request_method='POST')
    config.add_route('delete_key', '/keys/{key}', request_method='DELETE')
    config.add_route('get_messages', '/messages/{key}', request_method='GET')

    # Healthcheck views
    config.add_route("version", "/__version__", request_method="GET")
    config.add_route("heartbeat", "/__heartbeat__", request_method="GET")
    config.add_route("lbheartbeat", "/__lbheartbeat__", request_method="GET")

    config.scan(".views")
    return config.make_wsgi_app()


push_messages/views.py
import json

from pyramid.exceptions import NotFound
from pyramid.response import Response
from pyramid.view import view_config

import push_messages


@view_config(route_name='get_keys', renderer='json')
def get_keys(request):
    return dict(keys=[{'public-key': v} for v in
                      request.redis.hkeys("registered_keys")])


@view_config(route_name='register_key')
def register_key(request):
    public_key = request.swagger_data['key']['public-key']
    request.redis.hset("registered_keys", public_key, "")
    loc = '/'.join([request.host_url, "keys", public_key])
    return Response(status=201, headers={'Location': str(loc)})


@view_config(route_name='delete_key')
def delete_key(request):
    key = request.matchdict['key']
    request.redis.hdel("registered_keys", key)
    return Response(status=204)


@view_config(route_name='get_messages', renderer='json')
def get_messages(request):
    public_key = request.matchdict['key']
    if not request.redis.hexists("registered_keys", public_key):
        raise NotFound()

    messages = request.redis.lrange(public_key, 0, 200)
    messages = filter(None, messages)
    loaded_messages = [json.loads(message) for message in messages]
    if not messages:
        return Response(status=204)
    return {
        'messages': [
            {'id': m['id'],
             'timestamp': float(m['timestamp']),
             'size': int(m['size']),
             'ttl': int(m['ttl'])} for m in loaded_messages
        ]
    }


@view_config(context="pyramid.exceptions.NotFound")
def empty_404(request):
    return Response("", status_code=404)


# Health-check views

@view_config(route_name="version", renderer="json")
def version(request):
    return dict(
        source="https://github.com/mozilla-services/push-messages/",
        version=push_messages.__version__,
        commit="",
    )


@view_config(route_name="heartbeat", renderer="json")
def heartbeat(request):
    request.redis.lrange("dummykey", 0, 10)
    return {}


@view_config(route_name="lbheartbeat", renderer="json")
def lbheartbeat(request):
    return {}
